### PR TITLE
[kendo menu] Make it possible to use custom renderers (i.e. Wicket co…

### PR DESCRIPTION
**Pull request by martin-g**
…mponents) for menu items

### CODE REVIEW

**Comment by martin-g** [comment]
There are some duplications in the samples' code because of the need to keep the code in the page (java & markup).

**Comment by sebfz1** [comment]
Hi Martin, I will take of this asap, thanks!

File: wicket-jquery-ui-samples/src/main/java/com/googlecode/wicket/jquery/ui/samples/pages/kendo/menu/KendoContextMenuPage.html [hunk 1]
```diff
@@ -13,6 +13,12 @@
 	<div wicket:id="menu"></div>
 	<br/>
 	<div wicket:id="feedback"></div>
+
+	<wicket:fragment wicket:id="moveToPosition">
+		<form wicket:id="form" onclick="var $evt=jQuery.event.fix(event);if($evt.target.nodeName !== 'BUTTON'){$evt.stopPropagation()}">
```
Comment (sebfz1) [inline wicket-jquery-ui-samples/src/main/java/com/googlecode/wicket/jquery/ui/samples/pages/kendo/menu/KendoContextMenuPage.html]: - I understand the purpose of this, but I think it is not straightforward for the user willing to achieve the same usecase... This is just a thought but, what about a special Form which can handle this (ie: Menu.InnerForm, MenuForm, etc)
- `jQuery.event.fix`. In the sample you can use `$` (just fyi)

File: wicket-jquery-ui-samples/src/main/java/com/googlecode/wicket/jquery/ui/samples/pages/kendo/menu/KendoContextMenuPage.html [hunk 2]
```diff
@@ -13,6 +13,12 @@
...
+			A menu item with a Form: <input wicket:id="position" type="text" style="width: 40px"/> <button type="button" class="k-button" wicket:id="moveToPositionBtn">Set position</button>
```
Comment (sebfz1) [inline wicket-jquery-ui-samples/src/main/java/com/googlecode/wicket/jquery/ui/samples/pages/kendo/menu/KendoContextMenuPage.html]: - is `class="k-button"` needed?

File: wicket-jquery-ui-samples/src/main/java/com/googlecode/wicket/jquery/ui/samples/pages/kendo/menu/KendoMenuPage.java [hunk 3]
```diff
@@ -74,11 +84,76 @@ public KendoMenuPage()
...
+		 * @param id The component id
+		 * @param menuItem
+		 */
+		public MoveToPositionPanel(String id, KendoMenuPage.MoveToPositionMenuItem menuItem, MarkupContainer markupProvider) {
+			super(id, "moveToPosition", markupProvider);
+
+			Form<Void> form = new Form<Void>("form");
+			add(form);
+
+			final TextField<Integer> position = new TextField<Integer>("position", new PropertyModel<Integer>(menuItem, "position"));
```
Comment (sebfz1) [inline wicket-jquery-ui-samples/src/main/java/com/googlecode/wicket/jquery/ui/samples/pages/kendo/menu/KendoMenuPage.java]: What about a nice `NumberTextField`? :p

File: wicket-jquery-ui-samples/src/main/java/com/googlecode/wicket/jquery/ui/samples/pages/kendo/menu/KendoContextMenuPage.html (hunk 1 above)
Comment (martin-g) [inline wicket-jquery-ui-samples/src/main/java/com/googlecode/wicket/jquery/ui/samples/pages/kendo/menu/KendoContextMenuPage.html]: Generally I agree with you. But I'm not sure you can provide a generic solution that will work for everyone.

Comment (martin-g) [inline wicket-jquery-ui-samples/src/main/java/com/googlecode/wicket/jquery/ui/samples/pages/kendo/menu/KendoContextMenuPage.html]: But yes, we can provide such form and users can extend from it. If someone needs more sophisticated solution then (s)he can use his/her own Form.

File: wicket-jquery-ui-samples/src/main/java/com/googlecode/wicket/jquery/ui/samples/pages/kendo/menu/KendoContextMenuPage.html (hunk 2 above)
Comment (martin-g) [inline wicket-jquery-ui-samples/src/main/java/com/googlecode/wicket/jquery/ui/samples/pages/kendo/menu/KendoContextMenuPage.html]: Not needed. In my app I use Wicket's AjaxButton and I just style it.

File: wicket-jquery-ui-samples/src/main/java/com/googlecode/wicket/jquery/ui/samples/pages/kendo/menu/KendoMenuPage.java (hunk 3 above)
Comment (martin-g) [inline wicket-jquery-ui-samples/src/main/java/com/googlecode/wicket/jquery/ui/samples/pages/kendo/menu/KendoMenuPage.java]: Should be fine for the demo! ;-)

Comment (martin-g) [inline wicket-jquery-ui-samples/src/main/java/com/googlecode/wicket/jquery/ui/samples/pages/kendo/menu/KendoMenuPage.java]: I've even used it in my project!
Thanks for the suggestion!

**Comment by sebfz1** [comment]
Hi Martin,

As I am looking at this PR... do you mind if I rename Menu##refresh to #reload? Basically, I always used a #refresh method, but in the menu case it destroys the widget beforehand, so I think #reload is more accurate here. Thus, in the future, nothing prevents having a widget component providing both method... WDYT?

Thanks,
Sebastien.

**Comment by sebfz1** [comment]
Something like this:

```java
// public or private?
public void destroy(AjaxRequestTarget target)
{
	this.widgetBehavior.destroy(target);
}

// public or private?
public void refresh(AjaxRequestTarget target)
{
	target.add(this);
}

public void reload(AjaxRequestTarget target)
{
	this.destroy(target);
	this.refresh(target);
}
```

**Comment by martin-g** [comment]
:+1:

**Comment by sebfz1** [comment]
Hi Martin,

I've moved `MenuItemForm` to a `menu.form` package, and I've added `MenuItemAjaxButton` to ease usage as bubbling is required in this case.
Also, I've renamed `#refresh` to `#reload`, but `#refresh` also still exists, leading to a silent api break, be sure to change your call... :)

Thanks again,
Sebastien.

**Comment by martin-g** [comment]
Everything works just fine! :-)

**Comment by sebfz1** [comment]
:smile::tada:

**Comment by martin-g** [comment]
It seems `Menu#destroy(AjaxRequestTarget)` is not enough/full.
http://docs.telerik.com/kendo-ui/api/javascript/ui/contextmenu#methods-destroy states `Important: This method does not remove the ContextMenu element from DOM.`
In my app I need to repaint a parent container with Ajax and both ContextMenu and DropDownList do not remove their DOM nodes. This leads to DOM element leaks.

**Pull request merged**